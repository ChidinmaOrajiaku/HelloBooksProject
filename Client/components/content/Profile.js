import React from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { store } from '../../index';
import { postProfileRequest, updatePasswordRequest, updateProfileRequest } from '../../actions/profileAction';
import { putBookRequest } from '../../actions/booksAction';

class Profile extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      userId: (store.getState()).auth.user.id,
      // sort out
      profileResponse: (store.getState()).userState[0],
      image: '',
      description: '',
      status: '',
      interest: '',
      password: '',
      gender: '',
      data: [],
      profile: [],
      books: [],
      popular: [],
      success: false,
      errors: false
    }
    this.handleChange = this.handleChange.bind(this);
    this.onProfileSubmit = this.onProfileSubmit.bind(this);
    this.onPasswordSubmit = this.onPasswordSubmit.bind(this);
    this.onReturnBooks = this.onReturnBooks.bind(this);
    this.onProfileUpdate = this.onProfileUpdate.bind(this);
  }

  handleChange(event) {
    this.setState({[event.target.id]: event.target.value})
  }

  onProfileSubmit(event) {
    event.preventDefault();
    this.props.postProfileRequest(this.state.userId, this.state).then(
      () => {Materialize.toast(this.state.profileResponse, 2000, 'teal rounded')},
      (errors) =>{
        Materialize.toast(errors.response.data.message, 2000, 'red accent-3 rounded')
     this.setState({ errors: errors.response.data.message  })
      }
    )
  }

  onPasswordSubmit(event) {
    event.preventDefault();
    this.props.updatePasswordRequest(this.state.userId, this.state).then(
      () => { 
        Materialize.toast(this.state.profileResponse, 2000, 'teal rounded')
      },
      (errors) =>{
        Materialize.toast(errors.response.data.message, 2000, 'red accent-3 rounded')
     this.setState({ errors: errors.response.data.message  })
      }
    )
    console.log((store.getState()).userState[0])
  }

  onProfileUpdate(event) {
    event.preventDefault();
    this.props.updateProfileRequest(this.state.userId, this.state).then(
      () => { 
        Materialize.toast(this.state.profileResponse, 2000, 'teal rounded')
      },
      (errors) =>{
        Materialize.toast(errors.response.data.message, 2000, 'red accent-3 rounded')
     this.setState({ errors: errors.response.data.message  })
      }
    )
  }

  onReturnBooks(event) {
    event.preventDefault();
    this.props.putBookRequest(this.state.userId, event.target.value).then(
      () => { 
        Materialize.toast(this.state.profileResponse, 2000, 'teal rounded')
      },
      (errors) =>{
        Materialize.toast(errors.response.data.message, 2000, 'red accent-3 rounded')
     this.setState({ errors: errors.response.data.message  })
      }
    )
  }

  componentWillMount() {
    axios.get('/api/v1/users/' + this.state.userId).then((res) => {
      localStorage.getItem('jwtToken');
      this.setState({ data: res.data})
    });
    axios.get('/api/v1/users/' + this.state.userId + '/profile').then((res) => {
      localStorage.getItem('jwtToken');
      this.setState({profile: res.data})
    });
    axios.get('/api/v1/users/' + this.state.userId + '/books').then((res) => {
      localStorage.getItem('jwtToken');
      this.setState({books: res.data})
    });
    axios.get('/api/v1/users/books').then((res) => {
      localStorage.getItem('jwtToken');
      this.setState({ popular: res.data})
    });
  }
  
  componentDidMount() {
    $('.modal').modal({
      dismissible: true, 
      opacity: .5, 
      inDuration: 300, 
      outDuration: 200,
      startingTop: '4%', 
      endingTop: '10%',
    });
  } 

  render() {
    
    $(document).ready(function(){
      $('.tooltipped').tooltip({delay: 50});
      $('.scrollspy').scrollSpy()
    });

    const { postProfileRequest, updatePasswordRequest, putBookRequest } = this.props
    const user = this.state.data
    const { errors, success } = this.state
  return (
    <div className="profile">
      {this.state.profile.map ((profile, index) =>
    <div className="row"  key={index}>
      <div className="col s4 ">
        <div className="card">
          <h1 className="name"> Welcome {user.firstname}</h1>
            <div className="card-image">
              <img src={profile.image}/>
            </div>
            <div className="card-content">
            <p> Name: {user.firstname + ' ' + user.lastname} </p>
            <p> Username: { user.username }</p>
            <p> Email: { user.email} </p>
          </div>
          <a className="waves-effect waves-light btn modal-trigger" href="#modal1">Create Profile</a>
          <a className="waves-effect waves-light btn modal-trigger" href="#modal2">Change Password</a>
        </div>
        </div>
        <div className="col s2 ">
           <p className="equal"> => </p>
        </div>
        <div className="col s5">
           <p className="profileData"> Description </p>
           <p className="data"> {profile.description}</p>
           <p className="profileData"> Interest </p>
           <p className="data"> {profile.interest}</p>
           <p className="profileData"> Status </p>
           <p className="data"> {profile.status}</p>
           <p className="profileData"> Gender </p>
           <p className="data"> {profile.gender}</p>
           <a className="waves-effect waves-light btn modal-trigger button3" href="#modal3">Edit Profile</a>
        </div>
        </div>
          )
          }
          <div className="row getPopularBooks">
            <hr/>
            <h1 className="return"> Popular Books </h1>
          {this.state.popular.map ((books, index) =>
               <div className="col s3 " key={index}>
                  <div className="card hoverable">
                     <div className="card-image">
                         <img id="image" value={ books.image } onChange={ this.handleChange } src= { books.image }/>
                     </div>
                  </div>
             </div>   
          )
          }
          </div>
          <div className="row getUnreadBooks">
            <hr/>
            <h1 className="return"> Unreturned Books </h1>
          {this.state.books.map ((books, index) =>
               <div className="col s3 " key={index}>
                  <div className="card hoverable">
                     <div className="card-image">
                         <img id="image" value={ books.image } onChange={ this.handleChange } src= { books.image }/>
                         <button value= { books.booksId } className="btn tooltipped btn-floating halfway-fab waves-effect waves-light teal" data-position="bottom" data-delay="50" data-tooltip="Hi! Click to return" onClick= { this.onReturnBooks } type="submit" name="action" ><i className="material-icons">add</i></button>
                     </div>
                  </div>
             </div>   
          )
          }
          </div>
        <div id="modal1" className="modal">
        <div className="modal-content">
        <div className="row">
        <div className="col s12">
        <form onSubmit={this.onProfileSubmit} id="form">
        <div className="row">
           <div className="input-field col s12">
             <input value={this.state.image} onChange={this.handleChange} id="image" required="required" type="text" className="validate"/>
               <label htmlFor="image">Image</label>
          </div>
        </div>
        <div className="row">
           <div className="input-field col s12">
             <textarea value={this.state.description} onChange={this.handleChange} id="description" required="required" type="text" className="materialize-textarea"></textarea>
              <label htmlFor="description">Description</label>
          </div>
        </div>
        <div className="row">
          <div className="input-field col s12">
            <input value={this.state.status} onChange={this.handleChange} id="status" required="required" type="text" className="validate"/>
             <label htmlFor="email">Status</label>
          </div>
        </div>
        <div className="row">
          <div className="input-field col s12">
            <input value={this.state.interest} onChange={this.handleChange} id="interest" required="required" type="text" className="validate"/>
             <label htmlFor="email">Interest</label>
          </div>
        </div>
        <div className="row">
          <div className="input-field col s12">
            <input value={this.state.gender} onChange={this.handleChange} id="gender" required="required" type="text" className="validate"/>
             <label htmlFor="gender">Gender</label>
          </div>
        </div>
        <button className="btn waves-effect waves-light" type="submit" name="action">Create<i className="material-icons right">send</i></button>
      </form>
        </div>
      </div>
      </div>
      </div>
      <div id="modal2" className="modal">
        <div className="modal-content">
          <div className="row">
            <div className="col s12">
              <form onSubmit={this.onPasswordSubmit} id="form">
                <div className="row">
                  <div className="input-field col s12">
                    <input value={this.state.password} onChange={this.handleChange} id="password" required="required" type="password" className="validate"/>
                     <label htmlFor="password">Password</label>
                 </div>
               </div>
                <button className="btn waves-effect waves-light" type="submit" name="action">Update<i className="material-icons right">send</i></button>
             </form>
           </div>
         </div>
       </div>
     </div>
     <div id="modal3" className="modal">
        <div className="modal-content">
        <div className="row">
        <div className="col s12">
        <form onSubmit={this.onProfileUpdate} id="form">
        <div className="row">
           <div className="input-field col s12">
             <textarea value={this.state.description} onChange={this.handleChange} id="description" type="text" className="materialize-textarea"></textarea>
              <label htmlFor="description">Description</label>
          </div>
        </div>
        <div className="row">
          <div className="input-field col s12">
            <input value={this.state.status} onChange={this.handleChange} id="status" type="text" className="validate"/>
             <label htmlFor="email">Status</label>
          </div>
        </div>
        <div className="row">
          <div className="input-field col s12">
            <input value={this.state.interest} onChange={this.handleChange} id="interest"  type="text" className="validate"/>
             <label htmlFor="email">Interest</label>
          </div>
        </div>
        <div className="row">
          <div className="input-field col s12">
            <input value={this.state.gender} onChange={this.handleChange} id="gender" type="text" className="validate"/>
             <label htmlFor="gender">Gender</label>
          </div>
        </div>
        <button className="btn waves-effect waves-light" type="submit" name="action">Edit<i className="material-icons right">send</i></button>
      </form>
        </div>
      </div>
      </div>
      </div>
      </div>
  );
}
};

Profile.propTypes = {
  postProfileRequest: PropTypes.func.isRequired,
  updatePasswordRequest: PropTypes.func.isRequired,
  putBookRequest: PropTypes.func.isRequired,
  updateProfileRequest: PropTypes.func.isRequired
}

export default connect(null, { postProfileRequest, updatePasswordRequest, putBookRequest, updateProfileRequest }) (Profile);
