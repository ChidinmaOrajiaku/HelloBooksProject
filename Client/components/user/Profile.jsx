import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { getUserDataRequest } from '../../actions/getUser';
import { updatePassword } from '../../actions/updatePassword';

/**
 * @class Profile
 */
export class Profile extends React.Component {
  /**
     * @constructor
     * @param {object} props
     */
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      firstname: '',
      lastname: '',
      username: '',
      email: '',
      password: '',
      verifyPassword: '',
      errors: false,
      message: this.props.passwordUpdate,
    };
    this.handleChange = this.handleChange.bind(this);
    this.handlePassword = this.handlePassword.bind(this);
    this.handleReset = this.handleReset.bind(this);
  }

  /**
   * @returns {object} response object
   * @memberof Profile
   */
  componentDidMount() {
    this.props.getUserDataRequest(this.props.usersId);
    $('.modal').modal({
      dismissible: true,
      opacity: 0.5,
      inDuration: 300,
      outDuration: 200,
      startingTop: '2%',
      endingTop: '20%',
    });
  }

  /**
   * Receives next props and sets it in state
   * @returns {object} response object
   * @param {object} nextProps
   * @memberof Profile
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.getUserData) {
      this.setState({
        firstname: nextProps.getUserData.firstname,
        lastname: nextProps.getUserData.lastname,
        email: nextProps.getUserData.email,
        username: nextProps.getUserData.username
      });
    }
  }

  /**
 * Handles input field change and sets the state
 * @returns {object} response object
 * @param {object} event
 * @memberof Profile
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }

  /**
 * Resets password input field on close or submission
 * @param {object} event
 * @memberof Profile
 * @returns {object} response object
 */
  handleReset(event) {
    event.preventDefault();
    this.setState({
      password: '',
      verifyPassword: ''
    });
  }

  /**
   * Updates user password
   * @returns {object} response object
   * @param {object} event
   * @memberof Profile
   */
  handlePassword(event) {
    event.preventDefault();
    this.props.updatePassword(this.props.usersId, this.state);
    setTimeout(() => {
      if (this.props.passwordUpdate.isUpdated === true) {
        Materialize.toast('Successfully Updated', 2000, 'teal rounded');
        $('.modal').modal('close');
        this.setState({
          password: '',
          verifyPassword: ''
        });
      } else if (this.state.password === '') {
        Materialize.toast(this.props.passwordUpdate.error.response.data.message, 2000, 'red rounded');
      } else if (this.props.passwordUpdate.isUpdated === false) {
        Materialize.toast(this.props.passwordUpdate.error.response.data.message, 2000, 'red rounded');
      }
    }, 1000);
  }

  /**
     * React Element Markup
     * @returns {object} response object
     * @memberof Profile
     */
  render() {
    return (
      <div className="profile row">
        <div className=""> <NavigationBar /> </div>
        <h4 className="col s12 m9 offset-m2"> PROFILE DETAILS</h4>
        <hr className="col s12 m3 offset-m5"></hr>
        <div className="row">
          <div className="col s12 m3 offset-m5 profileB">
            <div className="card">
              <div className="card-content">
                <p className="profileDetails">
                  Name: {`${this.state.firstname} ${this.state.lastname}`}</p>
                <p className="profileDetails"> Username: {this.state.username}</p>
                <p className="profileDetails"> Email: {this.state.email}</p>
                <p className="profileDetails"> Password: ***
                  <a className="modal-trigger" href="#modal1">
                    <button className="material-icons" name="action">create</button>
                  </a>
                </p>
              </div>
            </div>
          </div>
        </div>
        <div id="modal1" className="modal">
          <div className="modal-content">
            <div className="row">
              <div className="input-field col s12">
                <input value={this.state.verifyPassword}
                  onChange={this.handleChange} id="verifyPassword" required="required"
                  type="password" className="validate"/>
                <label htmlFor="verifyPassword">Old Password</label>
              </div>
              <div className="input-field col s12">
                <input value={this.state.password}
                  onChange={this.handleChange} id="password" required="required"
                  type="password" className="validate"/>
                <label htmlFor="password">New Password</label>
              </div>
            </div>
          </div>
          <div className="modal-footer">
            <button onClick={this.handlePassword}
              className="passwordButton"
              id="passwordButton"> Change </button>
            <a className="modal-action modal-close">
              <button
              className="passwordCButton"
              onClick= {this.handleReset}
              id="passwordCButton">Cancel</button>
            </a>
          </div>
        </div>
        <div> <Footer /></div>
      </div>
    );
  }
}

export const mapStateToProps = state => (
  {
    usersId: state.auth.user.id,
    getUserData: state.getUser[0].response,
    passwordUpdate: state.updatePassword[0]
  }
);

export default connect(mapStateToProps, { getUserDataRequest, updatePassword })(Profile);
