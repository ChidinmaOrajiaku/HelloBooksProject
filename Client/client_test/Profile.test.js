import React from 'react';
import $ from 'jquery';
import toJson from 'enzyme-to-json';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Profile } from '../components/user/Profile.jsx';

configure({ adapter: new Adapter() });

describe('<Profile />', () => {
  const props = {
    passwordUpdate: jest.fn(),
    getUserDataRequest: jest.fn()
  };
  let mountedComponent;
  const ProfileItem = () => {
    if (!mountedComponent) {
      mountedComponent = shallow(<Profile {...props}/>);
    }
    return mountedComponent;
  };

  it('should match snapshot test', () => {
    const component = ProfileItem();
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
