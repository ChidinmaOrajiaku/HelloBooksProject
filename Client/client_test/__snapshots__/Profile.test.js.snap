// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> should match snapshot test 1`] = `
<div
  className="profile row"
>
  <div
    className=""
  >
     
    <Connect(withRouter(NavigationBar)) />
     
  </div>
  <h4
    className="col m9 offset-m2"
  >
     PROFILE DETAILS
  </h4>
  <hr
    className="col m3 offset-m5"
  />
  <div
    className="row"
  >
    <div
      className="col m3 offset-m5 profileB"
    >
      <div
        className="card"
      >
        <div
          className="card-content"
        >
          <p
            className="profileDetails"
          >
            Name: 
             
          </p>
          <p
            className="profileDetails"
          >
             Username: 
          </p>
          <p
            className="profileDetails"
          >
             Email: 
          </p>
          <p
            className="profileDetails"
          >
             Password: ***
            <a
              className="modal-trigger"
              href="#modal1"
            >
              <button
                className="material-icons"
                name="action"
              >
                create
              </button>
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="modal"
    id="modal1"
  >
    <div
      className="modal-content"
    >
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="verifyPassword"
            onChange={[Function]}
            required="required"
            type="password"
            value=""
          />
          <label
            htmlFor="verifyPassword"
          >
            Old Password
          </label>
        </div>
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="password"
            onChange={[Function]}
            required="required"
            type="password"
            value=""
          />
          <label
            htmlFor="password"
          >
            New Password
          </label>
        </div>
      </div>
    </div>
    <div
      className="modal-footer"
    >
      <button
        className="passwordButton"
        id="passwordButton"
        onClick={[Function]}
      >
         Change 
      </button>
      <a
        className="modal-action modal-close"
      >
        <button
          className="passwordCButton"
          id="passwordCButton"
        >
          Cancel
        </button>
      </a>
    </div>
  </div>
  <div>
     
    <Component />
  </div>
</div>
`;
